pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTaskRestletResource.java

```diff
@@ -149,6 +149,28 @@ public Map<String, TaskState> getTaskStatesByTable(
     return _pinotHelixTaskResourceManager.getTaskStatesByTable(taskType, tableNameWithType);
   }
 
+  @GET
+  @Path("/tasks/{taskType}/{tableNameWithType}/metadata")
+  @ApiOperation("Get task metadata for the given task type and table")
+  public String getTaskMetadataByTable(
+      @ApiParam(value = "Task type", required = true) @PathParam("taskType") String taskType,
+      @ApiParam(value = "Table name with type", required = true) @PathParam("tableNameWithType")
+          String tableNameWithType) {
+    return _pinotHelixTaskResourceManager.getTaskMetadataByTable(taskType, tableNameWithType);
+  }
+
+  @DELETE
```

Comment:
helix deletes task metadata after sometime, right ? Why provide an endpoint to delete it? Have you verified with helix that this ok? and there is no race condition?

Reply:
Helix task framework is not aware of the task metadata here, e.g. RealtimToOfflineTask keeps a timestamp as watermark in a ZNode under `/Propertystore/MINION_TASK_METADATA/` to manage its progress.